design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+## Design
+
+### Serialization
+
+If a construct is said to be serializable, it must implement the `ISerializable` interface:
+
+```ts
+interface ISerializable {
+  serialize(ctx: ISerializationContext): void;
+}
+```
+
+When an object is serialized, the `serialize` method is called with an object which implements the following interface:
+
+```ts
+interface ISerializationContext {
+  writeString(key: string, value: string, options?: SerializationOptions): void
+  writeObject(key: string, obj: ISerializable, options?: SerializationOptions): void;
+}
+
+interface SerializationOptions {
+  description?: string;
+}
+```
+
+The serialization context allows the object serialize itself through key/value strings via calls to `writeString(key, value)`. If the object encapsulates another serializable object, it can use `writeObject(key, obj)`, which will result in a subsequent call to the sub-object's `serialize` method with an appropriate context.
+
+### Deserialization
+
+To support deserialization, classes must also include a public static `deserializeXxx` method which reads the object from a deserialization context and returns an object that implements the resource type interface:
+
+> The reason we indicate the type in the method name is because static methods in JavaScript are inherited, so we can differentiate between `ExtraBucket.deserializeExtraBucket` and `ExtraBucket.deserizlizeBucket`.
+
+```ts
+class MyResource extends Construct implements ISerializable {
+  static deserializeMyResource(ctx: IDeserializationContext): IMyResource;
+}
+```
+
+The deserialization context is an object that implements the following interface:
+
+```ts
+interface IDeserializationContext {
+  scope: Construct;
+  id: string;
+  readString(key: string, options?: ReadStringOptions): string;
+  readObject(key: string): IDeserializationContext;
+}
+
+interface ReadStringOptions {
+  allowUnresolved?: boolean; /** @default true */
+}
+```
+
+`readString(key)` can be used to read values stored by `writeString`. and `readObject(key)` returns a deserialization context for composite deserialization written via `writeObject`.
+
+The `allowUnresolved` option for `readString` can be used by constructs to indicate that returned value must be a resolved value (i.e. not a token). This implies, for example, that when importing this value, users cannot use the `resolveType: Deployment`  option (REQ8).
+
+Since `deserializeXxx` will need to create new construct objects, the deserialization context will supply a consistent `scope` and `id` which can be used to instantiate a construct object that represents this object. For example, `scope` can be mapped to the current `Stack` and `id` can be mapped to `exportName` which is ensured to be unique within the environment (and therefore, the current stack).
+
+Implementers of `deserializeXxx` should check if a construct with `id` already exists within `scope` and return it instead of instantiating and new object (REQ6).
```

Comment:
See - it's interesting this says *should*, where the requirement implies *must*.